pandas/tests/groupby/test_categorical.py

```diff
@@ -1669,3 +1669,27 @@ def test_categorical_transform():
     expected["status"] = expected["status"].astype(delivery_status_type)
 
     tm.assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize("func", ["first", "last"])
+def test_groupby_first_on_categorical_col_grouped_on_2_categoricals(func: str):
+    # GH 34951
+
+    cat = pd.Categorical([0, 0, 1, 1])
+    val = [0, 1, 1, 0]
+    df = pd.DataFrame({"a": cat, "b": cat, "c": val})
+
+    idx = pd.Categorical([0, 1])
+    idx = pd.MultiIndex.from_product([idx, idx], names=["a", "b"])
+    expected = {
+        "first": pd.Series([0, np.NaN, np.NaN, 1], idx, name="c"),
+        "last": pd.Series([1, np.NaN, np.NaN, 0], idx, name="c"),
+    }
+
+    df_grp = df.groupby(["a", "b"])
+    df_res = getattr(df_grp, func)()
+    tm.assert_frame_equal(df_res, expected[func].to_frame())
+
+    srs_grp = df_grp["c"]
+    srs_res = getattr(srs_grp, func)()
```

Comment:
can you use 

result = 
expected = 
tm.assert_frame_equal(result, expected)

much easier to read

Reply:
Done